# Superday

In [37]:
import pandas as pd

In [38]:
df = pd.read_csv('Pre-Super_Day_candidate_dataset__28candidate_29.csv')
print(df.shape)
df.head(5)

(100000, 14)


User ID  applications  \
0  00007820-89cb-4c1d-9940-eb270d605a35             1   
1  00012b55-514c-421e-9c76-3300abbc1134             1   
2  000157c1-b6a3-4c86-82c7-9ec1bda3799a             1   
3  00020400-efab-4b10-8812-2a0aaf774841             1   
4  0002f737-0cda-48fb-91ed-533f3d0eab05             1   

                     Reason  Loan_Amount  FICO_score Fico_Score_group  \
0  cover_an_unexpected_cost       100000         669             fair   
1   credit_card_refinancing        70000         594             fair   
2          home_improvement        10000         596             fair   
3          home_improvement       100000         642             fair   
4            major_purchase        30000         642             fair   

  Employment_Status       Employment_Sector  Monthly_Gross_Income  \
0         full_time  consumer_discretionary                  5024   
1         full_time  information_technology                  5764   
2         full_time  information_technology                  4017   
3         part_time                  energy                  3129   
4         full_time                  energy                  4220   

   Monthly_Housing_Payment  Ever_Bankrupt_or_Foreclose Lender  Approved  \
0                      927                           0      B         0   
1                     1177                           0      B         0   
2                     1487                           0      A         0   
3                      904                           0      A         0   
4                     1620                           0      A         0   

   bounty  
0       0  
1       0  
2       0  
3       0  
4       0

## Feature Engineering

In [ ]:
from sklearn.preprocessing import LabelEncoder

In [39]:
df['Fico_Score_group'].unique()

array(['fair', 'poor', 'good', 'very_good', 'excellent'], dtype=object)

In [40]:
df['Employment_Status'].unique()

array(['full_time', 'part_time', 'unemployed'], dtype=object)

In [47]:
# Encode ordinal categorical variables in order and non-ordinal with one hot encoding 

fico_group_map = { 'poor': 0, 'fair': 1, 'good': 2, 'very_good': 3, 'excellent': 4 }
employment_status_map = { 'unemployed': 0, 'part_time': 1, 'full_time': 2 }

df_encoded = df.copy()
df_encoded = pd.get_dummies(df_encoded, columns=['Reason'], prefix=['Reason'])
df_encoded = pd.get_dummies(df_encoded, columns=['Employment_Sector'], prefix=['Employment_Sector'])
df_encoded["Fico_Score_group"] = df_encoded["Fico_Score_group"].map(fico_group_map)
df_encoded["Employment_Status"] = df_encoded["Employment_Status"].map(employment_status_map)
df_encoded.head(2)

User ID  applications  Loan_Amount  \
0  00007820-89cb-4c1d-9940-eb270d605a35             1       100000   
1  00012b55-514c-421e-9c76-3300abbc1134             1        70000   

   FICO_score  Fico_Score_group  Employment_Status  Monthly_Gross_Income  \
0         669                 1                  2                  5024   
1         594                 1                  2                  5764   

   Monthly_Housing_Payment  Ever_Bankrupt_or_Foreclose Lender  ...  \
0                      927                           0      B  ...   
1                     1177                           0      B  ...   

   Employment_Sector_consumer_discretionary  \
0                                      True   
1                                     False   

   Employment_Sector_consumer_staples  Employment_Sector_energy  \
0                               False                     False   
1                               False                     False   

   Employment_Sector_financials  Employment_Sector_health_care  \
0                         False                          False   
1                         False                          False   

   Employment_Sector_industrials  Employment_Sector_information_technology  \
0                          False                                     False   
1                          False                                      True   

   Employment_Sector_materials  Employment_Sector_real_estate  \
0                        False                          False   
1                        False                          False   

   Employment_Sector_utilities  
0                        False  
1                        False  

[2 rows x 29 columns]

## Tell us about the variables
_Possible things to consider: Which variables are the most helpful in understanding if a customer is going to be approved or denied for a loan? Are there certain variables that are not useful to collect?_

In [29]:
from sklearn.ensemble import RandomForestClassifier

In [30]:
features = df.drop('Lender', axis=1)  # Features
target = df['Lender']
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)
rf_classifier.fit(features, target)

ValueError: could not convert string to float: '00007820-89cb-4c1d-9940-eb270d605a35'

## Tell us about the lenders
_Possible things to consider: What is each Lender’s average approval rate? How does their revenue per approval differ? Are there any clear differences between the three different lenders on what type of customers they approve?_

Percent% approval by Vendor

In [9]:
df[['Lender', 'Approved']].groupby('Lender').mean() * 100

Approved
Lender           
A       10.965455
B        7.127273
C       17.057143

In [24]:
approved = df[df['Approved'] == 1]

In [25]:
approved[(approved['Lender'] == 'A')].describe()

applications    Loan_Amount   FICO_score  Monthly_Gross_Income  \
count        6031.0    6031.000000  6031.000000           6031.000000   
mean            1.0   42417.509534   697.358481           7506.998674   
std             0.0   28226.467212    74.113488           3389.702307   
min             1.0    5000.000000   353.000000           2007.000000   
25%             1.0   20000.000000   659.500000           4857.500000   
50%             1.0   30000.000000   702.000000           7272.000000   
75%             1.0   70000.000000   739.000000           9612.000000   
max             1.0  100000.000000   850.000000          19997.000000   

       Monthly_Housing_Payment  Ever_Bankrupt_or_Foreclose  Approved  bounty  
count              6031.000000                 6031.000000    6031.0  6031.0  
mean               1697.039794                    0.005472       1.0   250.0  
std                 670.735658                    0.073775       0.0     0.0  
min                 300.000000                    0.000000       1.0   250.0  
25%                1263.000000                    0.000000       1.0   250.0  
50%                1724.000000                    0.000000       1.0   250.0  
75%                2109.500000                    0.000000       1.0   250.0  
max                3299.000000                    1.000000       1.0   250.0

In [27]:
approved[(approved['Lender'] == 'B')].describe()

applications    Loan_Amount   FICO_score  Monthly_Gross_Income  \
count        1960.0    1960.000000  1960.000000           1960.000000   
mean            1.0   41790.816327   732.108673           8053.576020   
std             0.0   27723.422857    53.841987           3256.772291   
min             1.0    5000.000000   600.000000           2014.000000   
25%             1.0   20000.000000   690.000000           5569.500000   
50%             1.0   30000.000000   724.000000           7891.500000   
75%             1.0   60000.000000   772.000000          10016.000000   
max             1.0  100000.000000   850.000000          19909.000000   

       Monthly_Housing_Payment  Ever_Bankrupt_or_Foreclose  Approved  bounty  
count              1960.000000                      1960.0    1960.0  1960.0  
mean               1694.571939                         0.0       1.0   350.0  
std                 667.931901                         0.0       0.0     0.0  
min                 300.000000                         0.0       1.0   350.0  
25%                1248.000000                         0.0       1.0   350.0  
50%                1710.500000                         0.0       1.0   350.0  
75%                2109.250000                         0.0       1.0   350.0  
max                3287.000000                         0.0       1.0   350.0

In [28]:
approved[(approved['Lender'] == 'C')].describe()

applications    Loan_Amount   FICO_score  Monthly_Gross_Income  \
count        2985.0    2985.000000  2985.000000           2985.000000   
mean            1.0   41135.678392   674.770519           6322.161809   
std             0.0   27824.071795    81.548255           3137.702364   
min             1.0    5000.000000   358.000000           2008.000000   
25%             1.0   20000.000000   622.000000           3936.000000   
50%             1.0   30000.000000   682.000000           5634.000000   
75%             1.0   60000.000000   724.000000           8253.000000   
max             1.0  100000.000000   850.000000          19982.000000   

       Monthly_Housing_Payment  Ever_Bankrupt_or_Foreclose  Approved  bounty  
count              2985.000000                 2985.000000    2985.0  2985.0  
mean               1344.649581                    0.016080       1.0   150.0  
std                 672.804328                    0.125806       0.0     0.0  
min                 301.000000                    0.000000       1.0   150.0  
25%                 791.000000                    0.000000       1.0   150.0  
50%                1260.000000                    0.000000       1.0   150.0  
75%                1875.000000                    0.000000       1.0   150.0  
max                3297.000000                    1.000000       1.0   150.0

### Takeaways
- Lender C is the most likely to approve applications at 17% approval, more than twice as much as vendor B at 7%
- Lender B targets the 'Prime' sector, while Lender C targets the lowest earners and Lender A splits the difference. Approved applicants for Lender B have an average income of `8053` per month, compared to Lender C at `6322`. Average Fico score for lender B is `732` compared to C's average of `674`.
- Lender B never approves a person with a foreclosure
- The average loan amount approved by each vendor is roughly the same at ~`$41,000`
- Lender B pays the most for each conversion at `$350`, followed by Lender A at `$250` and Lender C at `150`


Lender B pays the most for each conversion, but also only approves "highest quality" candidate. Lender C approves more people but also pays less. Lender A splits the difference

## Tell us about which customers we should match to each lender